### ISSUE TITLE
Examples in Documentation

### ISSUE THREAD

**Issue by BedirT** [applegrew/django-select2#525]
I would love to see some examples in the documentation. I want to use the classes in a complex form system, and I am a Django beginner, so it's really hard to figure out everything without anything to look up to.

**Comment by codingjoe** [comment]
Hi @BedirT,

there are at least to some degree examples. To be fair, most of them are in the API documentation. Could you give me a bit more context what it is you are looking for exactly. I believe good open source projects thrive with good documentation, so I am very eager to improve it.

Best
-Joe

**Comment by BedirT** [comment]
Hi @codingjoe,

Well, what I have is a combination of a couple system together.

One of my original class was;

```python
class questionMethodForm(forms.ModelForm):

	class Meta:
		methodQuery = QMethod.objects.all()
		mthChoices = (('other', 'other'),)
		for obj in methodQuery:
			mthChoices = mthChoices + ((str(obj.method), str(obj.method)),)
		model = QMethod
		fields = ['method']
		widgets = {
			'method': forms.CheckboxSelectMultiple(attrs={'class':'form-control'}, choices=tuple(set(mthChoices))),
		}
```

It has only one field, which I used ```forms.CheckboxSelectMultiple``` (as can be seen). Well I was trying to adapt select2's ```ModelSelect2TagWidget```, which clearly states that [we need to implement our own function](
https://django-select2.readthedocs.io/en/latest/django_select2.html#django_select2.forms.ModelSelect2TagWidget)

In my case, the model I am using is connected to another via an m2m connection. So I couldn't put the points together, and adapt the example given to my own situation.

## models.py
```python
class QMethod(models.Model):	
    questions = models.ManyToManyField(Question)
    method = models.CharField(max_length=40)

    def __str__(self):
        return self.method
```

It would be really helpful if there was a clear explanation for all the widgets one by one, I am sure that people are already using them in their applications, so it shouldn't be too hard to gather them (It's just an assumption, I am not aware of it if it is).

**Comment by BedirT** [comment]
Well, I still couldn't solve my issue, but I fully understood and commented the function as simply as I could. If anyone else is not that advanced in Django, here is the fixed and working version of the function MyModelSelect2TagWidget as well as explained in detail.

```python
class MyModelSelect2TagWidget(ModelSelect2TagWidget):
	queryset = myModel.objects.all()

	def value_from_datadict(self, data, files, name):
		values = super().value_from_datadict(data, files, name)
		'''
		values now stores the data entered in  the format
		Entered: 301, 222
		If there is already 301 with the id '59'.

		['59', '222']

		'''

		qs = self.queryset.filter(**{'pk__in': list(values)})
		'''
		qs now stores the objects from the queryset which database
		already has them included.
		i.e. if there was 301 already in the db.
		qs: <QuerySet [<QLevel: 301>]> 

		You can get the data by using this object - i.e. qs[0].pk
		'''

		pks = set(str(o.pk) for o in qs)
		'''
		Now pks stores the set of the primary keys that are
		inputted by user but also exist in the database.
		i.e.
		{'59'}
		'''
		cleaned_values = []
		'''
		We are using cleaned_values to move the data next step
		'''

		for val in values:
			if str(val) not in pks:
				val = self.queryset.create(level=val).pk
				'''
				- self.queryset.create(level=val): 
					part is to add the new data to the db
                                        here level is a key we have in the model
					i.e. 222
				- val = ....pk:
					part is to store the new data's primary key
					so that we will add it to cleaned_values and
					will return that result at the end
				'''
			cleaned_values.append(val)

		return cleaned_values
```

**Comment by codingjoe** [comment]
Hi @BedirT sorry for the late response. I must admit, I still don't understand what you are trying to do. Let's start with the form, what's your form doing? What special about it? How do you want to use select2 to change it's behavior? Did you manage to make it work without select2?

To be frank, django-select2 is only a wrapper for select2. If you can't make it work without this library, I am not sure I can help you making it work at all.

**Comment by codingjoe** [comment]
Since this seems to be a very specialized question, I will close the ticket, feel free to further comment and ask questions. I will continue to answer, I just like to keep the tickets clean.

**Comment by BedirT** [comment]
Hi,

I actually changed my method but still having problems, i opened another issue (u can close that one too if needed) [Here](https://github.com/applegrew/django-select2/issues/527#issue-410958971). I couldn't solve the issue with including models(modelform) so I am even okay with using forms and creating the queries myself afterwards.

The thing is simply put, my form is created from 4 sections. I am using another app called [extra-views](https://django-extra-views.readthedocs.io/en/latest/) to combine them. I will give exact code example trying as simple as put:

So 4 models I have:
## Models.py
```python
class Part1(models.Model):
    year = models.CharField(max_length=5)
    source = models.CharField(max_length=140)
    author = models.ForeignKey(User, on_delete=models.PROTECT)

class Part2(models.Model):
    qID = models.ForeignKey(Part1, on_delete=models.CASCADE)
    level = models.CharField(max_length=16)

class Part3(models.Model):
    qID = models.ForeignKey(Part1, on_delete=models.CASCADE)
    method = models.CharField(max_length=16)

class Part4(models.Model):
    qID = models.ForeignKey(Part1, on_delete=models.CASCADE)
    subject = models.CharField(max_length=16)
```

And I have the form file that i create them and use select2 library. Here I tried using the custom ```value_from_datadict``` as I wrote above. But it creates multiple entries in database. So I tried it with simple ```','.join``` method. Let me show both down below:

## Forms.py
```python
# FORM v2
class YearCustomTitleWidget(ModelSelect2Widget):
	model = Question
	search_fields = ['year__icontains']
	
	def label_from_instance(self, obj):
		return force_text(obj).upper()

class QuestionSelect2TagWidget(YearCustomTitleWidget, Select2TagWidget):

	def value_from_datadict(self, data, files, name):
	    values = super(QuestionSelect2TagWidget, self).value_from_datadict(data, files, name)
	    return ",".join(values)

class questionForm(forms.ModelForm):
	class Meta:
		model = Question
		exclude = ('date_posted', 'author',)
		widgets = {
			'year': QuestionSelect2TagWidget(
				attrs={'data-placeholder': 'Choose the year of the question', 
				'data-width': '100%'},),
			'source': QuestionSelect2TagWidget(
				attrs={'data-placeholder': 'Choose the question source', 
				'data-width': '100%'})
               }

# FORM v1
class MyModelSelect2TagWidget(ModelSelect2TagWidget):
	queryset = QLevel.objects.all()

	def value_from_datadict(self, data, files, name):
		values = super().value_from_datadict(data, files, name)
		qs = self.queryset.filter(**{'pk__in': list(values)})
		pks = set(str(o.pk) for o in qs)
		cleaned_values = []
		for val in values:
			if str(val) not in pks:
				val = self.queryset.create(level=val).pk
			cleaned_values.append(val)

		return cleaned_values

class questionLevelForm(forms.ModelForm):
	class Meta:
		model = QLevel
		fields = ('level',)
		widgets = {
			'level': MyModelSelect2TagWidget(
				attrs={'data-placeholder': 'Enter the question levels', 
					'data-width': '100%'},
				queryset=QLevel.objects.all(),
				search_fields=['level__icontains'],),
		}

	def __init__(self, *args, **kwargs):
		super(questionLevelForm, self).__init__(*args, **kwargs)
```
**The problem with v1 is:**
   - The db is getting multiple entries (I will explain in more detail), it is related to ```val = self.queryset.create(level=val).pk``` and getting ```cleaned_values```. Only cleaned values would work, but then I can't get rid of entries that already exist in db...

**The problem with v2 is:**
   - The db is getting multiple entries again + This time I can't use searching. 
   - Trying this method only because I want to get the data and process it myself (manual query) 

#### Multiple entries?
I want to get i.e. level i.e. 301 and then 302 in next row. What I get is 301 and 302 and next row is [301,302] if i use v1. If i use v2 I get only [301,302].

Please help, I literally asked everywhere. If i cant work it out i might have to change all the project, and start over with js.

**Comment by codingjoe** [comment]
Hi @BedirT,

I am happy to help as much as I can. Can you explain this a bit more: I don't understand the difference in this statement:

> I want to get i.e. level i.e. 301 and then 302 in next row. What I get is 301 and 302 and next row is [301,302] if i use v1. If i use v2 I get only [301,302].

Also can you maybe create a small sample app and send me the code here via a zip file? That would make things a bit simpler.

**Comment by BedirT** [comment]
Hi @codingjoe,

Try this form, and check the admin page for the db's you will hopefully understand the issue.

[trySelect2.zip](https://github.com/applegrew/django-select2/files/2895743/trySelect2.zip)

THANKS!

**Comment by codingjoe** [comment]
Hi @BedirT you didn't answer my question yet. I must say, I am having trouble understanding what it is you want from me. Please be very specific and lets try to focus on one problem at a time.

**Comment by BedirT** [comment]
Alright let me try again. 

If you tried the app I sent, (running on redis), using a superadmin (for the sake of ease) basically we have this form:

<img width="465" alt="screen shot 2019-02-25 at 1 19 17 pm" src="https://user-images.githubusercontent.com/10860127/53362716-6b084c80-3900-11e9-98de-8e5effec8e43.png">

I am filling it out like this for example:

<img width="366" alt="screen shot 2019-02-25 at 1 20 20 pm" src="https://user-images.githubusercontent.com/10860127/53362747-7bb8c280-3900-11e9-970c-5de984bbaa80.png">

When we press the button and go to admin page, in method table its like this:

<img width="451" alt="screen shot 2019-02-25 at 1 26 17 pm" src="https://user-images.githubusercontent.com/10860127/53363125-5aa4a180-3901-11e9-9559-2f7df2c2928b.png">

which the element is connected to the question with a foreign key:

<img width="514" alt="screen shot 2019-02-25 at 1 26 34 pm" src="https://user-images.githubusercontent.com/10860127/53363311-c38c1980-3901-11e9-91dc-907ed8c55642.png">

although what i want is this:

<img width="271" alt="screen shot 2019-02-25 at 1 31 12 pm" src="https://user-images.githubusercontent.com/10860127/53363322-c8e96400-3901-11e9-9ac7-23b6e4658e20.png">

<img width="496" alt="screen shot 2019-02-25 at 1 31 18 pm" src="https://user-images.githubusercontent.com/10860127/53363329-cc7ceb00-3901-11e9-9395-68dc7bdf8319.png">

I am trying to overcome this with the custom ```value_from_datadict``` in the *level* part but there still seems to be another problem:

<img width="488" alt="screen shot 2019-02-25 at 1 27 39 pm" src="https://user-images.githubusercontent.com/10860127/53363382-efa79a80-3901-11e9-892c-711a87dc553d.png">

Multiple objects are okay - with different foreign keys - but the part with ```[12, 13]``` which is the ids of the already existing ones. So what I am trying to achieve is:

<img width="321" alt="screen shot 2019-02-25 at 1 36 29 pm" src="https://user-images.githubusercontent.com/10860127/53363541-604eb700-3902-11e9-91df-fe1e0fa4b186.png">

<img width="515" alt="screen shot 2019-02-25 at 1 35 48 pm" src="https://user-images.githubusercontent.com/10860127/53363549-647ad480-3902-11e9-82fb-8fe26a6ab318.png">
<img width="528" alt="screen shot 2019-02-25 at 1 36 21 pm" src="https://user-images.githubusercontent.com/10860127/53363550-647ad480-3902-11e9-8e54-4cd74506e8eb.png">

**Comment by codingjoe** [comment]
Ok, here's what I see.
In form 2 you need to drop the `return ','.join(values)` part. This creates the `Math,Angles,Geometry` issue.

In form 1 you need to use `get_or_create` and make the label unique. The PK check only works if the value provided was a PK and therefore from an existing object.

I'd really recommend digging though the testapp and to read the example documented for each widget.
